File: src/main/java/dev/hoot/api/game/GameThread.java
package dev.hoot.api.game;

import lombok.extern.slf4j.Slf4j;
import meteor.Main;
import meteor.rs.ClientThread;

import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.FutureTask;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

@Slf4j
public class GameThread
{
	private static final long TIMEOUT = 1000;

	public static void invoke(Runnable runnable)
	{
		if (Main.INSTANCE.getClient().isClientThread())
		{
			runnable.run();
		}
		else
		{
			ClientThread.INSTANCE.invokeLater(runnable);
		}
	}

	public static <T> T invokeLater(Callable<T> callable)
	{
		if (Main.INSTANCE.getClient().isClientThread())
		{
			try
			{
				return callable.call();
			}
			catch (Exception e)
			{
				e.printStackTrace();
			}
		}

		try
		{
			FutureTask<T> futureTask = new FutureTask<>(callable);
			ClientThread.INSTANCE.invokeLater(futureTask);
			return futureTask.get(TIMEOUT, TimeUnit.MILLISECONDS);
		}
		catch (ExecutionException | InterruptedException | TimeoutException e)
		{
			e.printStackTrace();
			throw new RuntimeException("Client thread invoke timed out after " + TIMEOUT + " ms");
		}
	}
}

File: src/main/kotlin/meteor/api/game/GameThread.kt
package meteor.api.game

import Main.client
import meteor.rs.ClientThread
import java.util.concurrent.*

object GameThread {

 var clientThread = ClientThread

    private const val timeout = 500L

    fun <T> invoke(func: () -> T) {
        try {
            if (client.isClientThread) {
                func()
            } else {
                clientThread.invokeLater(Runnable { func() })
            }
        } catch (e: Exception) {

        }
    }

    fun <T> invokeLater(func: () -> T): T? {
        if (client.isClientThread) {
            return try {
                func()
            } catch (e: Exception) {
                null
            }
        }

        return try {
            val futureTask = FutureTask(Callable(func))
            clientThread.invokeLater(futureTask)
            futureTask.get(timeout, TimeUnit.MILLISECONDS)
        } catch (e: Exception) {
            null
        }
    }
}